Add topojson support / EMS v3 support

thomasneirynck opened the pull request:
This PR adds support for Topojson and the EMS v3 maps backend. It also refactors some of the common code between Coordinate Map and Region Map.

---

~~This is a rough outline for adding topojson support to Region Maps. Closes https://github.com/elastic/kibana/issues/14331.~~
 
Adding topojson layers raises a few new issues that need to be solved.

- [x] I'm not sure the current single format as a string parameter will be sufficient. We need more metadata for topojson , which includes the path where the features are in the topology. 
- [x] the current joining mechanism is too slow. We potentially need to join tens of thousands of geojson features against thousands of ES term-buckets. ~~I changed it to use a map, but even that should be improved. It's probably wrong to rely on the onFeature callbacks coming out of Leaflet, which have no defined order. Rather, we should just do that join ourselves.~~ (SHOULD EVALUATE IF THIS WAS WORTH IT)
- [x] the current outline styling is too thick. Since we'll have more features, they'll tend to be smaller, so we should make the outline configurable (cf. https://github.com/elastic/kibana/issues/15360)
- [x] we should offer the ability to exclude shapes from the map if they don't match any of the terms. Right now, they stay on as unstyled features. The annoying part is that to do with we have to recreate the leaflet layer from scratch because filtering cannot be applied after layer instantiation.
- [x] multi-baselayer support. This requires more changes to the manifest, so we can display something human readable.

-----

nyurik commented:
In theory we could standardize our topojsons to always have the same object hierarchy. We don't have to use multiple top-level subobjects, but instead standardize on `{objects:{data: ...}}`. In this case we won't need to store extra topojson file-specific metadata.

nyurik commented:
P.S. I used this approach in Kartotherian - https://github.com/kartotherian/geoshapes/blob/master/geoshapes.js#L406

thomasneirynck commented:
I'm going to expand this PR to support other EMS v3 features
- include multi-baselayer support
- add docs for custom configs for on-prem deployments

thomasneirynck commented:
I added support for multibase layer (hypothetical, since EMS doesn't publish multiple layers yet). It's a rough outline, not intende to go in as-is. I changed the wms-options now to wrap the entire baselayer configuration.

For some reason, this broke the Coordinate Map visualization. It's broken even when excluding the wms-options Not sure why yet..

thomasneirynck commented:
@timroes @ppisljar This is an outline still, but most of the features that should go in, are in. I'm not sure if we need "more" features in order to support the use-case. Tests/docs are missing.

The biggest issue is that we actually don't have an EMS v3 yet. So this requires some manual configurations in the yaml to make topojson work.

Things can be evaluated already:
- do join upfront and don't rely on looping over features in Leaflet. This is a huge performance win when dealing with a lot of shapes. It does introduce more complexity though. For example, we need to rejoin each time the terms-data changes.
- ability to only display matching shapes. This has an impact on our ability of reusing choroplethlayer-instances. You can't remove features from leaflet-layers or change filters-on-the-fly, so we might have to recreate the instance if the data changes.
- multi-baselayer support (right now only showing single option, see EMSv3 comment above). I redid the existing wms-option directive and moved it there (this addition impacts the Coordinate Map as well).
- new line-width configuration

I'd like to merge https://github.com/elastic/kibana/pull/15507 as well, so I can rebase with the added unit tests. So I'd prioritize that one first for review.

timroes commented on src/core_plugins/tile_map/public/tile_map_vis.js:
```diff
@@ -61,9 +60,9 @@ VisTypesRegistryProvider.register(function TileMapVisType(Private, getAppState,
+        baseLayers: []
       },
-      optionsTemplate: tileMapTemplate,
+      optionsTemplate: '<tile_map-vis-params></tile_map-vis-params>',
```
`<tile-map-vis-params>`

timroes commented on src/core_plugins/tile_map/public/editors/tile_map_vis_params.js:
```diff
@@ -0,0 +1,13 @@
+  return {
+    restrict: 'E',
+    template: tileMapTemplate,
+    link: function () {
```
* Get rid of the `link` function if we don't need it.
* Use child scope for the directive instead of writing to `vis.params` directly in the template.

ppisljar replied:
+1

nyurik commented:
I wonder if we should integrate https://github.com/briannaAndCo/Leaflet.Antimeridian into this as well, to solve all potential anti-meridian issues. Might need to evaluate performance hit, or make it optional (e.g. the catalog file will mark a certain dataset as requiring a special antimeridian handling)

thomasneirynck commented:
Attached is an example topojson file that you can use for testing. It works with the makelogs data.

[world_countries.topojson.zip](https://github.com/elastic/kibana/files/1602364/world_countries.topojson.zip)

Drop it in `src/core_plugins/region_map/public/data` and use following configuration in the `kibana.yml`:
```
regionmap:
  includeElasticMapsService: true
  layers:
     - name: "World countries (self hosted, topojson)"
       url: "../plugins/region_map/data/world_countries.topojson"
       format:
          type: 'topojson'
       meta:
          feature_collection_path: 'collection'
       fields:
          - name: "iso2"
            description: "Two letter abbreviation"
```

ppisljar commented on package.json:
```diff
@@ -202,6 +202,7 @@
     "tabbable": "1.1.0",
     "tar": "2.2.0",
     "tinygradient": "0.3.0",
+    "topojson-client": "3.0.0",
```
The module doesn't seem to be much used and kept up to date ? (seems last change was half a year ago). Can we rely on it ?

thomasneirynck replied:
it seems legit to me. Written by the creator of d3 and frequently downloaded. It's a subset of `https://www.npmjs.com/package/topojson`, with only some conversion functions.

ppisljar commented on src/core_plugins/kibana/inject_vars.js:
```diff
@@ -10,7 +10,6 @@ export function injectVars(server) {
   const regionmapsConfig = serverConfig.get('regionmap');
   const mapConfig = serverConfig.get('map');
 
-
```
ups ?

thomasneirynck replied:
?

ppisljar replied:
the only change in this file is the removal of this empty line ....

thomasneirynck replied:
oh, oops! ;)

ppisljar commented on src/core_plugins/region_map/public/region_map_vis.js:
```diff
@@ -36,6 +36,8 @@ VisTypesRegistryProvider.register(function RegionMapProvider(Private, regionmaps
         wms: config.get('visualization:tileMap:WMSdefaults'),
         mapZoom: 2,
         mapCenter: [0, 0],
+        outlineWeight: 1,
```
what was the default outlineWeight before ?

thomasneirynck replied:
1

ppisljar commented on src/core_plugins/region_map/public/region_map_vis_params.js:
```diff
@@ -27,6 +27,11 @@ uiModules.get('kibana/region_map')
                 const layerFromService = layersFromService[i];
                 const alreadyAdded = newVectorLayers.some((layer) => layerFromService.layerId === layer.layerId);
                 if (!alreadyAdded) {
+                  if (layerFromService.format === 'geojson') {//backfill v1 manifest for now
```
the comment should probably move on its own line to make it easier to notice.

ppisljar commented on src/core_plugins/tile_map/public/base_maps_visualization.js:
```diff
@@ -88,6 +87,8 @@ export function BaseMapsVisualizationProvider(serviceSettings) {
       options.center = centerFromUIState ? centerFromUIState : this.vis.params.mapCenter;
 
       this._kibanaMap = new KibanaMap(this._container, options);
+      this._kibanaMap.setMinZoom(0);//use a default
```
why are this defaults hardcoded here and not read from vis.params ?

thomasneirynck replied:
These settings are driven by the base-layers and those are read from the EMS-manifests. These are just placeholders when we create the map, but they get overridden as baselayers are configured.

I did create variables for them.

ppisljar commented on src/core_plugins/tile_map/public/base_maps_visualization.js:
```diff
@@ -99,75 +100,62 @@ export function BaseMapsVisualizationProvider(serviceSettings) {
+      try {
+
+        if (mapParams.wms.enabled) {
+          const minZoom = 0;
```
should define on top and use just once, or even better use defaults set on vis.params ?

ppisljar commented on src/core_plugins/tile_map/public/base_maps_visualization.js:
```diff
@@ -99,75 +100,62 @@ export function BaseMapsVisualizationProvider(serviceSettings) {
+        if (mapParams.wms.enabled) {
+          const minZoom = 0;
+          const maxZoom = 18;
+          if (maxZoom > this._kibanaMap.getMaxZoomLevel()) {
```
why only updating map zoom if maxZoom is bigger than current maxZoomLevel ?

ppisljar commented on src/core_plugins/tile_map/public/coordinate_maps_visualization.js:
```diff
@@ -25,13 +25,13 @@ export function CoordinateMapsVisualizationProvider(Notifier, Private) {
 
 
     async _makeKibanaMap() {
-
```
changes to this file are unnecessary, should remove to make the change cleaner

ppisljar commented on src/core_plugins/region_map/public/region_map_visualization.js:
```diff
@@ -49,7 +49,12 @@ export function RegionMapsVisualizationProvider(Private, Notifier, config) {
       }
 
-      this._updateChoroplethLayer(this._vis.params.selectedLayer.url, this._vis.params.selectedLayer.attribution);
+      this._updateChoroplethLayerForNewMetrics(
```
can you quickly explain why all this calls to updateChoroplethLayer ? (4 calls to updateChoroplethLayerForXXX)

thomasneirynck replied:
After  looking into this more in-depth, I can't really trigger this being called 4 times in a single `render` call.
 
This may get called 2 times in a single render call. Once when there is a data change

These checks arose due to a limitation of leaflet, that doesn't allow us to change the data after we have assigned it to a layer. That's why we need to check if we can reuse the existing layer or create a new instance. Whether we can reuse it depends on a number of factors:
- if any parameters change that affect the join-parameters, we cannot reuse the instance
- if the data changes, we cannot reuse the instance.

I think it is working as intended. The checks avoid recreating an instance of the leaflet-layer and downloading new data when we shouldn't. This is an important feature. If we remove this and redraw the visualization from scratch for each render call, this will absolutely kill responsiveness of the map.

ppisljar commented on src/core_plugins/region_map/public/choropleth_layer.js:
```diff
@@ -181,85 +312,91 @@ export default class ChoroplethLayer extends KibanaMapLayer {
-      return boundsOfAllFeatures.isValid() ? boundsOfAllFeatures : null;
-    }
-  };
+function makeColorDarker(color) {
```
this diff got really bad, most of this functions did not really change just their position in the file did ....

thomasneirynck commented:
@nyurik wrt. antemeridian. I wouldn't make this part of this PR, but we can consider it. Our reliance on leaflet plugins is a long term liability rather than an asset. I'd reevaluate when we get more community feedback on this issue.

ppisljar commented:
i don't have any more comments on code and as far as i could test this seems to work.
waiting for docs and tests to pass.

timroes commented on src/core_plugins/tile_map/public/editors/wms_options.js:
```diff
@@ -2,10 +2,53 @@ import { uiModules } from 'ui/modules';
+    },
+    link: function ($scope) {
+
+      console.log('call linker wms options');
```
Needs to be removed

timroes commented on src/core_plugins/tile_map/public/editors/wms_options.js:
```diff
@@ -2,10 +2,53 @@ import { uiModules } from 'ui/modules';
+            if (!$scope.options.selectedTmsLayer) {
+              $scope.options.selectedTmsLayer = $scope.options.tmsLayers[0];
+            }
+            resolve(true);
```
I guess the `true` isn't needed here?

thomasneirynck replied:
I preferred the promise to resolve to a success-flag iso. of just undefined, as it's being used as a boolean check by the client.

timroes replied:
I am fine with leaving it in, even though I think, the promise should just reject if there are cases where it's not successful, and a resolved promise should always mean successful, which would render the `true` useless imho. But I won't nitpick on this one :D

ppisljar replied:
i agree 100% that if promise was not succesful it should reject instead of resolve with false

timroes commented on src/ui/public/vis/map/service_settings.js:
```diff
@@ -65,9 +71,13 @@ uiModules.get('kibana')
         this._loadFileLayers = _.once(async () => {
           const catalogue = await this._loadCatalogue();
-          const fileService = catalogue.services.filter((service) => service.type === 'file')[0];
+          const fileServices = catalogue.services.filter((service) => service.type === 'file');
```
I think this could be shortened, by using the `.find` method?

```js
const fileService = catalogue.services.find(service => service.type === 'file');
if (!fileService) {
  return [];
}
```

timroes commented on src/ui/public/vis/map/service_settings.js:
```diff
@@ -77,26 +87,23 @@ uiModules.get('kibana')
-          const firstService = _.cloneDeep(services[0]);
-          if (!firstService) {
-            throw new Error('Manifest response does not include sufficient service data.');
+          const tmsServices = catalogue.services.filter((service) => service.type === 'tms');
```
`.find` could be used here too

thomasneirynck commented:
jenkins test this

thomasneirynck commented:
"Tile map with no changes will update with visualization change"-dashboard test seems flaky.

This did not seem to fail here: https://kibana-ci.elastic.co/job/elastic+kibana+pull-request+multijob-selenium/10782/console 

Looking into it..

ppisljar commented:
jenkins, test this

thomasneirynck commented:
Will create tracker ticket for failing dashboard test Not sure why it is flaky, needs some investigation.

thomasneirynck commented:
jenkins, test this

timroes commented:
Please don't comment out tests, but rather use `.skip` so it will still be tracked, that these tests exist, and just skipped for the moment. Please also open an issue for all skipped tests, to investigate and get them working again.

ppisljar commented on src/core_plugins/tile_map/public/kibana_map.js:
```diff
@@ -398,7 +398,7 @@ export class KibanaMap extends EventEmitter {
     const northWest = bounds.getNorthWest();
     if (
-      southEast.lng === northWest.lng &&
+      southEast.lng === northWest.lng ||
```
why this change ?

thomasneirynck replied:
the and-case didn't trap the case where the visualization would have a width, but not a height (or vice-versa).

ppisljar commented:
seems old map visualizations (saved before this PR) no longer work on dashboard:

![screenshot-localhost-5601 2018-01-16 09-42-42-354](https://user-images.githubusercontent.com/13629809/34979518-c92632a8-faa1-11e7-8ec6-0e080cea0229.png)

timroes commented:
Does this PR break previously created visualizations, that has a custom map server configured. They would now require a EMSv3 server, while beforehand an EMSv2. Is EMSv3 backward compatible, so this visualization will still work with any EMSv2 server? If not I think we need a breaking change warning on this, since it would require all users with custom map servers to update Kibana in parallel to upgrade their map server to use EMSv3.

ppisljar commented:
old visualizations should still work, we can't expect users to recreate all their map visualizations.

same seems to happen for region map

thomasneirynck commented:
@timroes there doesn't need to be backward compatibility between Kibana 6.2 and up, and EMSv2. Kibana 6.2 will only connect to EMSv3 and never to any of the older endpoints. Likewise, none of the oler versions of Kibana will connect to EMSv3.

From the EMS end, we do need to ensure that every service that is published in v2 remains available in EMSv3. This is to ensure that  maps saved in older versions will have the same look&feel when reopened in a later version of Kibana.

thomasneirynck commented:
thanks @ppisljar . That's a bug due to the service-settings for EMS only being loaded in the UI-editor component. I will fix this.

thomasneirynck commented:
@timroes there are also no user-hosted EMS-versions, so users will not be able to run v2 themselves.

ppisljar commented on test/functional/apps/dashboard/_dashboard_time.js:
```diff
@@ -78,15 +78,15 @@ export default function ({ getPageObjects }) {
-        expect(prettyPrint).to.equal('Today');
-      });
-    });
+    // describe('time changes', function () {
```
what about this test? could we reenable this one as well ?

ppisljar commented:
also update PR description and open a new issue for the docs and link it here ?

thomasneirynck commented:
@ppisljar that test is flaky and fails/succeeds locally. it doesn't even have anything to do with maps, or even a populated dashboard. I created a separate ticket #16055, I'd prefer not get hung up on it for this PR.

+1 on separate doc ticket

thomasneirynck commented:
last test run before merging. 

jenkins, test this

thomasneirynck commented:
Backport: 
- [ ] https://github.com/elastic/kibana/pull/16076

bhavyarm commented:
@thomasneirynck do we have functional tests for this PR? Thanks!

thomasneirynck commented:
@bhavyarm there's no additional functional test, although there is more unit test.

This is mostly a backend change to accomodate the new EMS-service, less so a functional change in Kibana.
